crates/sui-bridge/src/action_executor.rs

```diff
@@ -0,0 +1,894 @@
+        certificate: VerifiedCertifiedBridgeAction,
+    ) {
+        let status = effects.status();
+        match status {
+            SuiExecutionStatus::Success => {
+                info!(?tx_digest, "Sui transaction executed successfully");
+                store
+                    .remove_pending_actions(&[certificate.data().digest()])
+                    .unwrap_or_else(|e| {
+                        panic!("Write to DB should not fail: {:?}", e);
+                    })
+            }
+            SuiExecutionStatus::Failure { error } => {
+                // In practice the transaction could fail because of running out of gas, but really
+                // should not be due to other reasons.
+                // This means manual intervention is needed. So we do not push them back to
+                // the execution queue because retries are mostly likely going to fail anyway.
+                // After human examination, the node should be restarted and fetch them from WAL.
+
+                // TODO metrics + alerts
+                error!(?tx_digest, "Manual intervention is needed. Sui transaction executed and failed with error: {error:?}");
+            }
+        }
+        *gas_object_ref = Self::refresh_gas_data_with_effects(sui_address, effects);
+    }
+
+    fn refresh_gas_data_with_effects(
+        sui_address: SuiAddress,
+        effects: SuiTransactionBlockEffects,
+    ) -> ObjectRef {
+        let updated_gas_object = effects.gas_object();
+        let obj_ref = updated_gas_object.reference.clone().to_object_ref();
+        // TODO: when we add multiple gas support in the future we could discard
+        // transferred gas object instead.
+        assert_eq!(
+            updated_gas_object.owner,
+            Owner::AddressOwner(sui_address),
+            "Gas object {:?} is no longer owned by address {}",
+            obj_ref.0,
+            sui_address
+        );
+        obj_ref
+    }
+
+    async fn refresh_gas_data_with_gas_object_id(
+        sui_address: SuiAddress,
+        gas_object_id: ObjectID,
+        sui_client: &SuiClient<C>,
+    ) -> ObjectRef {
+        let (gas_obj_ref, owner) = sui_client.get_gas_object_ref_and_owner(gas_object_id).await;
+
+        // TODO: when we add multiple gas support in the future we could discard
+        // transferred gas object instead.
+        assert_eq!(
+            owner,
+            Owner::AddressOwner(sui_address),
+            "Gas object {:?} is no longer owned by address {}",
+            gas_object_id,
+            sui_address
+        );
+        gas_obj_ref
+    }
+}
+
+pub async fn submit_to_executor(
+    tx: &mysten_metrics::metered_channel::Sender<BridgeActionExecutionWrapper>,
+    action: BridgeAction,
+) -> Result<(), BridgeError> {
+    tx.send(BridgeActionExecutionWrapper(action, 0))
+        .await
+        .map_err(|e| BridgeError::Generic(e.to_string()))
+}
+
+pub fn build_transaction(gas_object_ref: &ObjectRef) -> TransactionData {
```

Comment:
this is dummy for testing. Will be replaced with actual tx building once the function signature is settled on move.